docs/docs/contracts.md

```diff
@@ -0,0 +1,703 @@
+
+Record field updates are written `r{f = v}`. This creates a new record value
+which is the same as `r`, but with the value of the field `f` replaced by `v`.
+Similarly, `m{[k] = v}` constructs a map with the same values as `m` except
+that `k` maps to `v`. It makes no difference if `m` has a mapping for `k` or
+not.
+
+It is possible to give a name to the old value of a field or mapping in an
+update: instead of `acc{ balance = acc.balance + 100 }` it is possible to write
+`acc{ balance @ b = b + 100 }`, binding `b` to `acc.balance`. When giving a
+name to a map value (`m{ [k] @ x = v }`), the corresponding key must be present
+in the map or execution fails, but a default value can be provided:
+`m{ [k = default] @ x = v }`. In this case `x` is bound to `default` if
+`k` is not in the map.
+
+Updates can be nested:
+```
+function clear_history(a : address, accounts : map(address, account)) : map(address, account) =
+  accounts{ [a].history = [] }
+```
+This is equivalent to `accounts{ [a] @ acc = acc{ history = [] } }` and thus
+requires `a` to be present in the accounts map. To have `clear_history` create
+an account if `a` is not in the map you can write (given a function `empty_account`):
+```
+  accounts{ [a = empty_account()].history = [] }
+```
+
+#### Map implementation
+
+Internally in the VM maps are implemented as hash maps and support fast lookup
+and update. Large maps can be stored in the contract state and the size of the
+map does not contribute to the gas costs of a contract call reading or updating
+it.
+
+### Strings
+
+There is a builtin type `string`, which can be seen as an array of bytes.
+Strings can be compared for equality (`==`, `!=`), used as keys in maps and
+records, and used in builtin functions `String.length`, `String.concat` and
+the hash functions described below.
+
+Please refer to the `String` [library documentation](sophia_stdlib.md#String).
+
+### Byte arrays
+
+Byte arrays are fixed size arrays of 8-bit integers. They are described in hexadecimal system,
+for example the literal `#cafe` creates a two-element array of bytes `ca` (202) and `fe` (254)
+and thus is a value of type `bytes(2)`.
+
+Please refer to the `Bytes` [library documentation](sophia_stdlib.md#Bytes).
+
+
+### Cryptographic builtins
+
+Libraries [Crypto](sophia_stdlib.md#Crypto) and [String](sophia_stdlib.md#String) provide functions to 
+hash objects, verify signatures etc. The `hash` is a type alias for `bytes(32)`.
+
+#### AEVM note
+The hash functions in `String` hash strings interpreted as byte arrays, and
+the `Crypto` hash functions accept an element of any (first-order) type. The
+result is the hash of the binary encoding of the argument as [described
+below](#encoding-sophia-values-as-binaries). Note that this means that for `s :
+string`, `String.sha3(s)` and `Crypto.sha3(s)` will give different results on AEVM.
+
+
+### Authorization interface
+
+When a Generalized account is authorized, the authorization function needs
```

Comment:
Here a warning that Auth.tx isn't available until Iris is is missing.

Reply:
done.

Reply:
Only Auth.tx is coming in Iris, the rest is already there. But there are no users of this, so nevermind 😢

Reply:
@hanssv I am planning to test generalized accounts and provide some examples in the near future ;-) ... but I don't know when I will find the time to do so. will definitely play around with it because we also want to support the feature in the java sdk